pkg/datasource/data.go

```diff
@@ -0,0 +1,334 @@
+		}
+		nf := &field{
+			Name:         f.FieldName(),
+			Index:        uint32(len(ds.fields) + len(newFields)),
+			PayloadIndex: idx,
+		}
+		if s, ok := f.(StaticField); ok {
+			nf.Size = s.FieldSize()
+			nf.Offs = s.FieldOffset()
+		} else {
+			return nil, fmt.Errorf("field %q is not statically sized or does not implement StaticField", nf.Name)
+		}
+		if s, ok := f.(TypedField); ok {
+			nf.Type = uint32(s.FieldType())
+		}
+		if tagger, ok := f.(TaggedField); ok {
+			nf.Tags = tagger.FieldTags()
+		}
+		if s, ok := f.(FlaggedField); ok {
+			nf.Flags |= uint32(s.FieldFlags())
+		}
+		if s, ok := f.(ParentedField); ok {
+			parent := s.FieldParent()
+			if parent >= 0 {
+				nf.Parent = uint32(parent + parentOffset)                          // TODO: validate?
+				nf.Flags |= FieldFlagHasParent.Uint32() | FieldFlagHidden.Uint32() // default to hide subfields
+			}
+		}
+
+		if nf.Offs+nf.Size > size {
+			return nil, fmt.Errorf("field %q exceeds size of container (offs %d, size %d, container size %d)", nf.Name, nf.Offs, nf.Size, size)
+		}
+		newFields = append(newFields, nf)
+	}
+
+	ds.fields = append(ds.fields, newFields...)
+
+	for _, f := range newFields {
+		ds.fieldMap[f.Name] = f
+	}
+
+	ds.payloadCount++
+
+	return &fieldAccessor{ds: ds, f: &field{
+		PayloadIndex: idx,
+		Size:         size,
+	}}, nil
+}
+
+func (ds *dataSource) AddField(name string, opts ...FieldOption) (FieldAccessor, error) {
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	if _, ok := ds.fieldMap[name]; ok {
+		return nil, fmt.Errorf("field %q already exists", name)
+	}
+
+	nf := &field{
+		Name:  name,
+		Index: uint32(len(ds.fields)),
+		Type:  uint32(Slice),
+	}
+	for _, opt := range opts {
+		opt(nf)
+	}
+
+	// Reserve new payload for non-empty fields
+	if nf.Flags&FieldFlagEmpty.Uint32() != FieldFlagEmpty.Uint32() {
+		nf.PayloadIndex = ds.payloadCount
+		ds.payloadCount++
+	}
+
+	ds.fields = append(ds.fields, nf)
+	ds.fieldMap[nf.Name] = nf
+	return &fieldAccessor{ds: ds, f: nf}, nil
+}
+
+func (ds *dataSource) GetField(name string) FieldAccessor {
+	ds.lock.RLock()
+	defer ds.lock.RUnlock()
+
+	f, ok := ds.fieldMap[name]
+	if !ok {
+		return nil
+	}
+	return &fieldAccessor{ds: ds, f: f}
+}
+
+func (ds *dataSource) GetFieldsWithTag(tag ...string) []FieldAccessor {
+	ds.lock.RLock()
+	defer ds.lock.RUnlock()
+
+	res := make([]FieldAccessor, 0)
+	for _, f := range ds.fields {
+		for _, t := range tag {
+			if slices.Contains(f.Tags, t) {
+				res = append(res, &fieldAccessor{ds: ds, f: f})
+			}
+		}
+	}
+	return res
+}
+
+func (ds *dataSource) Subscribe(fn DataFunc, priority int) {
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	if fn == nil {
+		return
+	}
+	ds.subscriptions = append(ds.subscriptions, &subscription{
+		priority: priority,
+		fn:       fn,
+	})
+	sort.SliceStable(ds.subscriptions, func(i, j int) bool {
+		return ds.subscriptions[i].priority < ds.subscriptions[j].priority
+	})
```

Comment:
Also only on initialization - depending on how many listeners to a datasource are. I'd say typically 3-5 times per init. Since the array is already in a sorted state all the time, an additional sort should be also not that costly - and this way we avoid another `PostInit()` or something like that and thus it's IMHO easier to understand.